target/linux/realtek/patches-5.10/713-add-smbus-support-for-sfp.patch

```diff
@@ -0,0 +1,115 @@
+--- a/drivers/net/phy/sfp.c
++++ b/drivers/net/phy/sfp.c
+@@ -412,23 +412,73 @@
+ 	return ret == ARRAY_SIZE(msgs) ? len : 0;
+ }
+ 
++static int sfp_smbus_read(struct sfp *sfp, bool a2, u8 dev_addr, void *buf,
++			  size_t len)
++{
++	union i2c_smbus_data data;
++	u8 bus_addr = a2 ? 0x51 : 0x50, *val = buf;
++	int ret;
++
++	while (len > 0) {
++		ret = i2c_smbus_xfer(sfp->i2c, bus_addr, 0, I2C_SMBUS_READ,
++			     dev_addr, I2C_SMBUS_BYTE_DATA, &data) < 0;
++		if (ret) 
++			return -EIO;
++
++		*val = data.byte;
++
++		val++;
++		dev_addr++;
++		len--;
++	}
++
++	return val - (u8 *)buf;
++}
++
++static int sfp_smbus_write(struct sfp *sfp, bool a2, u8 dev_addr, void *buf,
++			   size_t len)
++{
++	union i2c_smbus_data data;
++	u8 bus_addr = a2 ? 0x51 : 0x50;
++	int ret;
++
++	memcpy(&data.byte, buf, len);
++
++	ret = i2c_smbus_xfer(sfp->i2c, bus_addr, 0, I2C_SMBUS_WRITE,
++			     dev_addr, I2C_SMBUS_BYTE_DATA, &data);
++
++	return ret;
++}
++
+ static int sfp_i2c_configure(struct sfp *sfp, struct i2c_adapter *i2c)
+ {
+ 	struct mii_bus *i2c_mii;
+ 	int ret;
+ 
+-	if (!i2c_check_functionality(i2c, I2C_FUNC_I2C))
++	if (i2c_check_functionality(i2c, I2C_FUNC_I2C)) {
++		sfp->read = sfp_i2c_read;
++		sfp->write = sfp_i2c_write;
++
++		i2c_mii = mdio_i2c_alloc(sfp->dev, i2c);
++		if (IS_ERR(i2c_mii))
++			return PTR_ERR(i2c_mii);
++
++		i2c_mii->name = "SFP I2C Bus";
++	} else if (i2c_check_functionality(i2c, I2C_FUNC_SMBUS_BYTE_DATA)) {
++		sfp->read = sfp_smbus_read;
++		sfp->write = sfp_smbus_write;
++
++		i2c_mii = mdio_i2c_alloc(sfp->dev, i2c);
++		if (IS_ERR(i2c_mii))
++			return PTR_ERR(i2c_mii);
++
++		i2c_mii->name = "SFP SMBus";
++	} else {
+ 		return -EINVAL;
++	}
+ 
+ 	sfp->i2c = i2c;
+-	sfp->read = sfp_i2c_read;
+-	sfp->write = sfp_i2c_write;
+-
+-	i2c_mii = mdio_i2c_alloc(sfp->dev, i2c);
+-	if (IS_ERR(i2c_mii))
+-		return PTR_ERR(i2c_mii);
+ 
+-	i2c_mii->name = "SFP I2C Bus";
+ 	i2c_mii->phy_mask = ~0;
+ 
+ 	ret = mdiobus_register(i2c_mii);
+--- a/drivers/net/mdio/mdio-i2c.c
++++ b/drivers/net/mdio/mdio-i2c.c
+@@ -35,6 +35,8 @@
+ 	u8 addr[3], data[2], *p;
+ 	int bus_addr, ret;
+ 
++	return 0xffff;
```

Comment:
While the rest of this patch looks fine and only misses a patch description, a change like this is clearly not acceptable upstream (and we will want to push this upstream asap in order to not having to backport and maintain tons of out-of-tree code).

Reply:
Yes, this removes functionality for the case where you have a true (i.e. not SMBus) I2C hardware and you want to talk MDIO over it to a PHY on an SFP module. For the Realtek SoCs this is not the case. When upstreaming, we would just leave this patch out. Someone else already came up with the solution to this problem, here: https://bootlin.com/blog/sfp-modules-on-a-board-running-linux/ and they have the complete solution, which I could not test at the time I implemented this because I was lacking an Ethernet PHY SFP module. But they did not upstream this, so I am not sure how easy it is. What is interesting is, that the SFP driver already explicitly takes the 16byte read limitations into account, which seem to be also present on I2C compatible hardware, not just SMBus hw. Once we have the SFP+ ports itself working, my plan was to revisit this. I already have a 10GBaseT SFP+ module lying around for just that purpose.